extensions/positron-python/src/client/positron/languageServerManager.ts

```diff
@@ -5,73 +5,145 @@
+                    return;
+                }
+
+                if (this._lastForegroundSessionId === sessionId) {
+                    // The foreground session has not changed.
+                    return;
+                }
+
+                // Get the foreground session.
+                const sessions = await getActivePythonSessions();
+                const foregroundSession = sessions.find((session) => session.metadata.sessionId === sessionId);
+                if (!foregroundSession) {
+                    // The foreground session is for another language.
+                    return;
+                }
+
+                // Update the last foreground session.
+                this._lastForegroundSessionId = sessionId;
+
+                // Update the Python path as expected by Pyright.
+                this.updatePythonPath(foregroundSession.runtimeMetadata.runtimePath);
+
+                // Activate the LSP for the foreground session.
+                await activateConsoleLsp(foregroundSession);
+            }),
+        );
+    }
+
+    private async updatePythonPath(pythonPath: string): Promise<void> {
+        // Pyright expects the Python path to be updated.
+        // This behavior only applies to workspaces; non-workspace editors do not update properly yet.
+        let folderUri: vscode.Uri | undefined;
+        let configTarget: vscode.ConfigurationTarget;
+
+        const { workspaceFolders } = this._workspaceService;
+
+        if (workspaceFolders === undefined || workspaceFolders.length === 0) {
+            folderUri = undefined;
+            configTarget = vscode.ConfigurationTarget.Global;
+        } else if (this._workspaceService.workspaceFile) {
+            folderUri = this._workspaceService.workspaceFile;
+            configTarget = vscode.ConfigurationTarget.Workspace;
+        } else {
+            folderUri = workspaceFolders[0].uri;
+            configTarget = vscode.ConfigurationTarget.WorkspaceFolder;
+        }
+
+        await this._pythonPathUpdaterService.updatePythonPath(pythonPath, configTarget, 'ui', folderUri);
+    }
+
+    private registerSession(session: PythonRuntimeSession): void {
+        this._disposables.push(
+            session.onDidChangeRuntimeState(async (state) => {
+                if (state === positron.RuntimeState.Ready) {
+                    // The session is ready, activate the LSP.
+                    if (session.metadata.sessionMode === positron.LanguageRuntimeSessionMode.Console) {
+                        // If no foreground session was set yet, set this as the foreground session.
+                        // This may happen e.g. after reloading the window with a non-foreground
+                        // Python console in a multilanguage workspace.
+                        if (!this._lastForegroundSessionId) {
+                            this._lastForegroundSessionId = session.metadata.sessionId;
+                        }
```

Comment:
This will also be the "normal" case for the first session that the user starts IIUC. You get the `Ready` state before the `onDidChangeForegroundSession` notification i think.